pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,664 @@
+	createHomeWorkspaceRBACResources                    func(ctx context.Context, userName string, homeWorkspace logicalcluster.Name) error
+	searchForReadyWorkspaceInLocalInformers             func(logicalClusterName logicalcluster.Name, isHome bool, userName string) (found bool, retryAfterSeconds int, checkError error)
+	tryToCreate                                         func(ctx context.Context, userName string, workspaceToCheck logicalcluster.Name, workspaceType tenancyv1alpha1.ClusterWorkspaceTypeName) (retryAfterSeconds int, createError error)
+}
+
+type homeWorkspaceHandler struct {
+	homeWorkspaceHandlerBuilder
+	homeWorkspaceFeatureLogic
+}
+
+func (b homeWorkspaceHandlerBuilder) build() *homeWorkspaceHandler {
+	h := &homeWorkspaceHandler{}
+	h.homeWorkspaceHandlerBuilder = b
+	h.homeWorkspaceFeatureLogic = homeWorkspaceFeatureLogic{
+		searchForHomeWorkspaceRBACResourcesInLocalInformers: func(logicalClusterName logicalcluster.Name, userName string) (found bool, err error) {
+			return searchForHomeWorkspaceRBACResourcesInLocalInformers(h, logicalClusterName, userName)
+		},
+		createHomeWorkspaceRBACResources: func(ctx context.Context, userName string, logicalClusterName logicalcluster.Name) error {
+			return createHomeWorkspaceRBACResources(h, ctx, userName, logicalClusterName)
+		},
+		searchForReadyWorkspaceInLocalInformers: func(logicalClusterName logicalcluster.Name, isHome bool, userName string) (found bool, retryAfterSeconds int, checkError error) {
+			return searchForReadyWorkspaceInLocalInformers(h, logicalClusterName, isHome, userName)
+		},
+		tryToCreate: func(ctx context.Context, userName string, logicalClusterName logicalcluster.Name, workspaceType tenancyv1alpha1.ClusterWorkspaceTypeName) (retryAfterSeconds int, createError error) {
+			return tryToCreate(h, ctx, userName, logicalClusterName, workspaceType)
+		},
+	}
+
+	return h
+}
+
+func (h *homeWorkspaceHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if !h.localInformers.synced() {
+		h.apiHandler.ServeHTTP(rw, req)
+		return
+	}
+
+	ctx := req.Context()
+	lcluster, err := request.ValidClusterFrom(ctx)
+	if err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	user, ok := request.UserFrom(ctx)
+	if !ok {
+		err := errors.New("No user in HomeWorkspaces filter !")
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	if sets.NewString(user.GetGroups()...).Has(kuser.SystemPrivilegedGroup) {
+		// If we are the system privileged group, it might be a call from the virtual workspace
+		// in which case we also search the user in the soft impersonation header.
+		if impersonated, err := softimpersonation.UserInfoFromRequestHeader(req); err != nil {
+			responsewriters.InternalError(rw, req, err)
+			return
+		} else if impersonated != nil {
+			user = impersonated
+		}
+	}
+	requestInfo, ok := request.RequestInfoFrom(ctx)
+	if !ok {
+		responsewriters.InternalError(rw, req, errors.New("No request Info"))
```

Comment:
Are we sending back `metav1.Status` for these errors?

Reply:
What would it bring ?
Why would it be more required here than [there](https://github.com/kcp-dev/kcp/blob/main/pkg/proxy/handler.go#L47) for example ?

Reply:
this is used everywhere in kube in similar situations. But I agree that a Status would be cleaner. I guess nobody cares because this is for cases that never happen (only when plumbing is wrong).